stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -107,29 +107,65 @@ end
 openblas_get_config() = strip(unsafe_string(ccall((@blasfunc(openblas_get_config), libblas), Ptr{UInt8}, () )))
 
 """
-    set_num_threads(n)
+    set_num_threads(n::Integer)
+    set_num_threads(::Nothing)
+
+Set the number of threads the BLAS library should use equal to `n::Integer`.
+If `nothing` is passed to this function, julia tries to figure out the optimial number of threads.
+The exact heuristic is an implementation detail.
 
-Set the number of threads the BLAS library should use.
+On exotic variants of `BLAS` this function can fail, which is indicated by returning `nothing`.
 """
 function set_num_threads(n::Integer)
     blas = vendor()
-    if blas === :openblas
-        return ccall((:openblas_set_num_threads, libblas), Cvoid, (Int32,), n)
-    elseif blas === :openblas64
-        return ccall((:openblas_set_num_threads64_, libblas), Cvoid, (Int32,), n)
+    if blas === :openblas || blas == :openblas64
+        return ccall((@blasfunc(openblas_set_num_threads), libblas), Cvoid, (Int32,), n)
     elseif blas === :mkl
         # MKL may let us set the number of threads in several ways
         return ccall((:MKL_Set_Num_Threads, libblas), Cvoid, (Cint,), n)
-    end
-
-    # OSX BLAS looks at an environment variable
-    @static if Sys.isapple()
+    elseif Sys.isapple()
+        # OSX BLAS looks at an environment variable
         ENV["VECLIB_MAXIMUM_THREADS"] = n
+    else
+        @warn "Failed to set number of BLAS threads." maxlog=1
     end
 
     return nothing
 end
 
+function set_num_threads(::Nothing)
+    n = max(1, Sys.CPU_THREADS ÷ 2)
+    set_num_threads(n)
+end
+
+"""
+    get_num_threads()
+
+Get the number of threads the BLAS library is using.
+
+On exotic variants of `BLAS` this function can fail, which is indicated by returning `nothing`.
+"""
+function get_num_threads()
+    blas = LinearAlgebra.BLAS.vendor()
+    if blas === :openblas || blas === :openblas64
+        return ccall((@blasfunc(openblas_get_num_threads), libblas), Cint, ())
+    elseif blas == :mkl
+        return ccall((:mkl_get_max_threads, libblas), Cint, ())
+    elseif Sys.isapple()
+        key = "VECLIB_MAXIMUM_THREADS"
+        s = get(ENV, key, "")
+        nt = Base.tryparse(Cint, s)
+        if nt === nothing
+            @warn "Failed to read environment variable $key" maxlog=1
+        else
+            return nt
+        end
+    end
+    ret = nothing
+    @warn "Could not get number of BLAS threads. Returning `$ret` instead." maxlog=1
+    return ret
```

Comment:
I think it's nice that the current code ties the implementation and the warning message. But I fear other core devs might not like "unnecessary" dynamism (as @OkonSamuel pointed out).  I think it might be safe to just do:

```suggestion
    @warn "Could not get number of BLAS threads. Returning `nothing` instead." maxlog=1
    return nothing
```

Reply:
If we go with returning `nothing`, should there be any warnings at all? 

I thought these two were complementary approaches, either return something clearly marking the failure to read how many threads, or else return an integer & warn that it may not be correct.

Reply:
I think a warning won't do any hurt.

Reply:
If this is used as

```julia
old = get_num_threads()
# ...
set_num_threads(old)
```

then the end-user (which may not be who writes the function) cannot notice something fishy is going on unless there is a warning.

Also, I'm guessing this code pass is rarely used. So, I guess it's OK to make things verbose until somebody complains.  It's also possible for a user to suppress the warning.

Reply:
But this is a performance optimisation, those tend to give up silently. (Although I wish more of them had `@debug` statements.) That said I agree it shouldn't be triggered often.

If there is to be a warning, just one is enough, not two.

Reply:
I think the question is if it is OK to consider changing number of threads as _pure_ optimization.  It changes the result slightly:

```julia
julia> A = randn(300, 300); B = randn(300, 300);

julia> LinearAlgebra.BLAS.set_num_threads(1)

julia> C1 = @btime $A * $B;
  1.112 ms (2 allocations: 703.20 KiB)

julia> LinearAlgebra.BLAS.set_num_threads(2)

julia> C2 = @btime $A * $B;
  650.866 μs (2 allocations: 703.20 KiB)

julia> extrema(C1 .- C2)
(-3.552713678800501e-14, 4.263256414560601e-14)
```

I think it is worth notifying the user that something unexpected is going on.

Reply:
Perhaps one warning when `set` fails, and nothing on `get`, is the right way to go?

Reply:
So I guess you mean to merge the two warnings for macOS? But, since this code path would be rarely exercised, I think it's better to go with simplicity than a nicely formatted warning.

Reply:
I also think we should not overengineer this rare case. About silent fails of performance optimization, I think it makes a difference if an API level function fails at its one job or if some "optimization detail" burried deep in a long algorithm fails.